### ISSUE TITLE
Add/clarify documentation re. task CE vs. no CE

### ISSUE THREAD

**Issue by cmeeren** [giraffe-fsharp/Giraffe#327]
I have, for a long time, been a bit confused regarding writing `HttpHandler`s with `task { }` vs. writing them "bare" without the `task` CE:

* There are apparently some important differences in behavior (e.g. eager evalution) that should be clearly documented, see e.g. #326
* The documentation doesn't seem to be 100% clear (with samples) on exactly what should be returned in task vs. non-task in order to short-circuit, invoke next, skip processing, etc.
  * I know there are some `ReturnFrom` overloads that Giraffe uses, but their usage does not seem to be explicitly called out, just used in samples demonstrating other things
  * Is it possible at all to return `None` outside of a task CE to skip processing, or is that task-only?
* Is there a performance impact to using `task` unnecessarily?

I can't contribute much, both due to time constraints and the fact that I don't fully understand the system (hence the questions), but I thought I'd take the time to post an issue with the question I could think of from the top of my head.

Sorry if I've missed something in the documentation.

**Comment by dustinmoris** [comment]
Thank you highlighting this gap in the documentation. I've tried to document this better now, which you can review here:
https://github.com/giraffe-fsharp/Giraffe/blob/develop/DOCUMENTATION.md#ways-of-creating-a-new-httphandler

Please let me know what you think?

**Comment by cmeeren** [comment]
Looks better. A couple of points:

```f#
"name"
|> ctx.TryGetQueryStringValue
|> defaultArg
<| "Giraffe"
```

can be written simpler (IMHO) as

```f#
ctx.TryGetQueryStringValue "name"
|> Option.defaultValue "Giraffe"
```

---

> In Giraffe there are three scenarios which a given HttpHandler can invoke:
>
> * Continue with next handler
> * Return early
> * Abort

I suggest using another term than **abort**, which either sounds like it does just the same as **return early**, or aborts the request altogether (whatever that may mean). Perhaps **ignore** or **skip** are better alternatives?

---

In **Return early**, this expression

```f#
(setStatusCode 401
>=> text "Please sign in.")
    skipRemainingPipeline ctx
```

seems like it's missing something (or the indentation is bad). The `HttpHandler` the first two lines evaluate to is ignored.

Furthermore, in this section, it would be nice to have a short repeat from earlier in the documentation that you can short circuit the execution by returning `Some HttpContext`.

Thanks for the good work :)

**Comment by dustinmoris** [comment]
```fsharp
ctx.TryGetQueryStringValue "name"
|> Option.defaultValue "Giraffe"
```

Not sure if this really works, did you try it? The `defaultArg` function requires a `'T option` as the first parameter but you are giving it a `string`?

> I suggest using another term than abort, which either sounds like it does just the same as return early, or aborts the request altogether (whatever that may mean). Perhaps ignore or skip are better alternatives?

Good point, I'll change it to skip as you suggested.

> ... seems like it's missing something (or the indentation is bad). The first two lines evaluate to an HttpHandler which is ignored.

Not sure I know what you mean. What do you mean it is missing something? Do you want me to just reformat the indentation or actually change the code?

> Furthermore, in this section, it would be nice to have a short repeat from earlier in the documentation that you can short circuit the execution by returning Some HttpContext.

Ok, I'll add it.

**Comment by cmeeren** [comment]
> ```fsharp
> ctx.TryGetQueryStringValue "name"
> |> Option.defaultValue "Giraffe"
> ```
> 
> Not sure if this really works, did you try it? The `defaultArg` function requires a `'T option` as the first parameter but you are giving it a `string`?

I'm using ‘Option.defaultValue‘, not ‘defaultArg‘. The operands are switched between the two; otherwise they're identical.

> Not sure I know what you mean. What do you mean it is missing something? Do you want me to just reformat the indentation or actually change the code?

Change the code. This is from the top of my head, but AFAIK, the first two lines evaluate to an HttpHandler. But the third line doesn't make use of it. I can't see that you forgot a pipe (the third line makes sense on its own), so I can't really say what these three lines are trying to do.

**Comment by dustinmoris** [comment]
> I'm using Option.defaultValue, not defaultArg.

Of course, sorry my bad! Yes that makes a lot more sense. I'm editing it right now :)

>  the first two lines evaluate to an HttpHandler. But the third line doesn't make use of it. I can't see that you forgot a pipe (the third line makes sense on its own), so I can't really say what these three lines are trying to do.

I think this is bad indentation on my part, which didn't make the example very clear.

Is this better?

```fsharp
let skipRemainingPipeline = Some >> Task.FromResult

let checkUserIsLoggedIn : HttpHandler =
    fun (next : HttpFunc) (ctx : HttpContext) ->
        if isNotNull ctx.User && ctx.User.Identity.IsAuthenticated
        then next ctx
        else
            (setStatusCode 401 >=> text "Please sign in.")
                skipRemainingPipeline ctx
```

In the `else` the first line is a handler which then needs `next` and `ctx` to be passed it, but in this instance instead of passing in `next` I am passing in `skipRemainingPipeline` in order to skip the `next` pipeline.

I have put the two parameters into a new line and indented so that it is not just one lone line, but I could also write it like this:

```fsharp
let skipRemainingPipeline = Some >> Task.FromResult

let checkUserIsLoggedIn : HttpHandler =
    fun (next : HttpFunc) (ctx : HttpContext) ->
        if isNotNull ctx.User && ctx.User.Identity.IsAuthenticated
        then next ctx
        else (setStatusCode 401 >=> text "Please sign in.") skipRemainingPipeline ctx
```

Which one do you prefer?

**Comment by cmeeren** [comment]
Regarding `skipRemainingPipeline`: I see, I didn't understand that it took the place of `next`. I thought `ctx` was applied to `skipRemainingPipeline`, not to the composed `HttpHandler` in front. I think your alternative 2 (which you have used in the documentation) is the most clear alternative, and the extra description you've added clarifies this further. I also suggest explicitly typing the function:

```f#
let skipRemainingPipeline : HttpFunc = Some >> Task.FromResult
```

You have already done this with `skip` further down, which is good. (Speaking of `skip`, there's an extra space in the definition that's not needed.)

---

This part could be clarified:

> Unless you define a `HttpHandler` function in the most verbose way with the `task {}` CE you are not executing any asynchronous operations until the handler gets invoked by the `GiraffeMiddleware`.
> 
> For example, in the example below, an `IDisposable` will get disposed **before** the actual `handler` gets executed:
> 
> ```f#
> let doSomething handler : HttpHandler =
>     fun next ctx ->
>         use __ = somethingToBeDisposedAtTheEndOfTheRequest
>         handler next ctx
> ```

The description says that no **asynchronous actions** are excecuted unless using the `task` overload, but the code (which is correct and demonstrates the problem) contains no asynchronous actions, so I think the problem is not directly related to asynchronous user code. if I'm correct, the description above the code block should be changed (though I'm not sure to what).

---

In the **Skip** section, you say that `route` is a good example, and then proceed to show not `route`, but `GET`. I think `GET` is the best and simplest example, but the text should be updated to match.

---

Nitpicking: Words starting with `Http` (`HttpHandler`, `HttpFunc`) should have the article "an", not "a".

**Comment by dustinmoris** [comment]
Great feedback, thank you! I'll address them shortly!

**Comment by dustinmoris** [comment]
> The use of the word "execute" makes it sound to me like it's saying that Task is lazily evaluated like Async, whereas my understanding is that the task (being eagerly evaluated) has started, it just hasn't necessarily completed unless you await it.

Correct and your phrasing is better than mine :). It was late for me. I'll submit this shortly! Thank you! This was very helpful, because the docs are really important and I think all your suggestions have massively improved them!
